File: grain/automation/transcript.py
"""Turns a captured Claude Code trajectory (docs/roadmap.md item 10; see
`capture.py`'s docstring for exactly what format this is and why) into a
small sequence of renderable events — pure parsing, no `curses`, so
`tui.py`'s actual screen-drawing code can stay a thin, mostly-untested
wrapper around this instead of the other way around. See `tui.py`'s own
docstring for why that split matters here specifically.

Each JSONL line is one Claude Code event, tagged by its own `type` field
(`system`, `user`, `assistant`, `result`, ...) — confirmed against a real
transcript, see `capture.py`. This module reads only the handful of fields a
trajectory viewer actually needs to render something useful; a field or
event type it doesn't recognize is shown plainly (its raw JSON), never an
error — Claude Code's own event shape grows new fields and subtypes across
versions, and this parser seeing one it doesn't know about should never make
a whole session unbrowsable.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class TranscriptEvent:
    role: str      # "user" | "assistant" | "system" | "result" | "raw" | ...
    summary: str    # one line, for the scrolling list view
    detail: str      # the fuller text, shown when this event is selected


def _text_of(content: Any) -> str:
    """A message's `content` is either a plain string (a typed-in user
    prompt) or a list of typed blocks (`text`, `thinking`, `tool_use`,
    `tool_result`) — both shapes appear in a real transcript (confirmed
    directly, see `capture.py`'s docstring).
    """
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    parts: list[str] = []
    for block in content:
        if not isinstance(block, dict):
            continue
        block_type = block.get("type")
        if block_type == "text":
            parts.append(block.get("text", ""))
        elif block_type == "tool_use":
            name = block.get("name", "tool")
            args = block.get("input", {})
            parts.append(f"[tool_use: {name} {json.dumps(args)}]")
        elif block_type == "tool_result":
            result_content = block.get("content", "")
            if isinstance(result_content, list):
                result_content = " ".join(
                    b.get("text", "") for b in result_content if isinstance(b, dict)
                )
            marker = "error" if block.get("is_error") else "ok"
            parts.append(f"[tool_result ({marker}): {result_content}]")
        elif block_type == "thinking":
            continue  # internal reasoning -- not shown in the trajectory view
    return "\n".join(p for p in parts if p)


def _one_line(text: str, limit: int = 100) -> str:
    text = " ".join(text.split())
    return text if len(text) <= limit else text[: limit - 1] + "…"


def parse_event(raw_line: str) -> TranscriptEvent | None:
    """One JSONL line -> one event, or None for a blank line (a trailing
    newline in the captured file, most commonly).
    """
    line = raw_line.strip()
    if not line:
        return None
    try:
        data = json.loads(line)
    except json.JSONDecodeError:
        # Not every byte on the wire is guaranteed valid JSON (a truncated
        # capture from a killed run, say) -- shown as-is rather than
        # dropped, so a malformed tail doesn't silently hide from the
        # viewer.
        return TranscriptEvent(role="raw", summary=_one_line(line), detail=line)

    event_type = data.get("type", "unknown")
    if event_type in ("user", "assistant"):
        message = data.get("message") or {}
        text = _text_of(message.get("content"))
        role = message.get("role", event_type)
        return TranscriptEvent(role=role, summary=f"{role}: {_one_line(text)}", detail=text)
    if event_type == "result":
        result_text = data.get("result", "") or ""
        cost = data.get("total_cost_usd")
        turns = data.get("num_turns")
        suffix = f" (cost=${cost}, turns={turns})" if cost is not None else ""
        return TranscriptEvent(
            role="result", summary=f"result:{suffix} {_one_line(result_text)}",
            detail=result_text or json.dumps(data, indent=2),
        )
    if event_type == "system":
        subtype = data.get("subtype", "")
        return TranscriptEvent(
            role="system", summary=f"system[{subtype}]",
            detail=json.dumps(data, indent=2),
        )
    return TranscriptEvent(
        role=str(event_type), summary=f"{event_type}", detail=json.dumps(data, indent=2),
    )


def parse_transcript(text: str) -> list[TranscriptEvent]:
    events = (parse_event(line) for line in text.splitlines())
    return [e for e in events if e is not None]

File: v2/pkg/agent/claude/transcript.go
package claude

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/bwsalmon/grain/v2/pkg/agent"
)

// rawEvent is one --output-format stream-json line -- the same event shape
// grain/automation/transcript.py parses (see its own docstring for the
// field-by-field justification: each line is one Claude Code event, tagged
// by its own "type" field), narrowed to only what this package needs to
// build an agent.Result rather than a full trajectory viewer.
type rawEvent struct {
	Type    string      `json:"type"`
	Subtype string      `json:"subtype"`
	IsError bool        `json:"is_error"`
	Message *rawMessage `json:"message"`
	Result  string      `json:"result"`
}

type rawMessage struct {
	Content json.RawMessage `json:"content"`
}

type rawBlock struct {
	Type string `json:"type"`
	Text string `json:"text"`
	// Thinking carries a "thinking"-type block's own text -- Claude's
	// extended-thinking output, present only when that feature is on, and
	// otherwise the one part of a real transcript neither FinalText nor
	// ToolCalls ever captured before bwsalmon/agents#446 gave it
	// somewhere to go (agent.Result.Transcript).
	Thinking  string          `json:"thinking"`
	Name      string          `json:"name"`
	Input     map[string]any  `json:"input"`
	ID        string          `json:"id"`
	ToolUseID string          `json:"tool_use_id"`
	Content   json.RawMessage `json:"content"`
	IsError   bool            `json:"is_error"`
}

// parseTranscript turns a full --output-format stream-json capture into a
// Result: FinalText from the terminal "result" event, ToolCalls built by
// pairing each "tool_use" block in an assistant message with the
// "tool_result" block bearing the same tool_use_id wherever it later
// appears in a user message -- the same pairing transcript.py's TUI does to
// render a trajectory, done here to produce agent.ToolCall values instead
// of display strings. A malformed or unrecognized line is skipped rather
// than treated as fatal, matching transcript.py's own tolerance: Claude
// Code's event shape grows new fields and subtypes across versions, and
// this parser seeing one it doesn't know about should not make a whole run
// unreadable.
//
// The same pass also builds Result.Transcript: every "thinking" and
// "text" block's own text, and a line for each tool call and what it got
// back, in the order they appeared -- the human-readable narrative
// FinalText and ToolCalls alone do not give a reader (bwsalmon/agents#446).
func parseTranscript(stdout string) (*agent.Result, error) {
	p := parseEvents(stdout)
	if p.resultErr != nil {
		return nil, p.resultErr
	}
	if !p.sawResult {
		return nil, fmt.Errorf("claude: no result event found in output")
	}
	p.result.Transcript = strings.TrimSpace(p.transcript.String())
	return &p.result, nil
}

// PartialTranscript renders whatever of a still-in-progress
// --output-format stream-json capture is on disk so far into the same
// human-readable narrative parseTranscript builds once a run finishes --
// what LiveTranscriptDir reads back for a run with no FinishedAt yet
// (bwsalmon/agents#467). Unlike parseTranscript it never errors: a
// missing (or not-yet-written) terminal "result" event just means the run
// is still going, not that anything is wrong, and a truncated final line
// -- reading a file the whole time claude is still appending to it can
// always catch mid-write -- is handled the same way parseTranscript
// already tolerates any other malformed line, by skipping it.
func PartialTranscript(stdout string) string {
	return strings.TrimSpace(parseEvents(stdout).transcript.String())
}

// parsedEvents is one line-by-line pass over a stream-json capture,
// shared by parseTranscript (the whole thing, once a run has finished)
// and PartialTranscript (however much exists so far) so the two can never
// build the transcript text itself two different ways.
type parsedEvents struct {
	result     agent.Result
	pending    map[string]int
	sawResult  bool
	resultErr  error
	transcript strings.Builder
}

func parseEvents(stdout string) *parsedEvents {
	p := &parsedEvents{pending: map[string]int{}}

	for _, line := range strings.Split(stdout, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var ev rawEvent
		if err := json.Unmarshal([]byte(line), &ev); err != nil {
			continue
		}

		switch ev.Type {
		case "assistant", "user":
			if ev.Message == nil {
				continue
			}
			for _, b := range parseBlocks(ev.Message.Content) {
				switch b.Type {
				case "thinking":
					if b.Thinking != "" {
						fmt.Fprintf(&p.transcript, "%s\n\n", b.Thinking)
					}
				case "text":
					if b.Text != "" {
						fmt.Fprintf(&p.transcript, "%s\n\n", b.Text)
					}
				case "tool_use":
					p.result.ToolCalls = append(p.result.ToolCalls, agent.ToolCall{Name: b.Name, Arguments: b.Input})
					p.pending[b.ID] = len(p.result.ToolCalls) - 1
					fmt.Fprintf(&p.transcript, "> %s(%s)\n", b.Name, inputSummary(b.Input))
				case "tool_result":
					if idx, ok := p.pending[b.ToolUseID]; ok {
						text := toolResultText(b.Content)
						p.result.ToolCalls[idx].Text = text
						p.result.ToolCalls[idx].IsError = b.IsError
						if b.IsError {
							fmt.Fprintf(&p.transcript, "! %s\n\n", text)
						} else {
							fmt.Fprintf(&p.transcript, "%s\n\n", text)
						}
					}
				}
			}
		case "result":
			p.sawResult = true
			if ev.IsError {
				p.resultErr = fmt.Errorf("claude: run ended in error (subtype=%s): %s", ev.Subtype, ev.Result)
				continue
			}
			p.result.FinalText = ev.Result
		}
	}
	return p
}

// inputSummary renders a tool_use block's own input as compact JSON for a
// transcript line -- best-effort, since a malformed value here should
// cost the transcript one unreadable line, never the whole parse (see
// parseTranscript's own doc comment on the same tolerance for the stream
// itself).
func inputSummary(input map[string]any) string {
	data, err := json.Marshal(input)
	if err != nil {
		return fmt.Sprintf("%v", input)
	}
	return string(data)
}

// parseBlocks reads a message's content field, which is either a plain
// string or a list of typed blocks (text, tool_use, tool_result) -- both
// shapes appear in a real transcript, per transcript.py's own _text_of.
func parseBlocks(raw json.RawMessage) []rawBlock {
	var blocks []rawBlock
	if err := json.Unmarshal(raw, &blocks); err == nil {
		return blocks
	}
	var s string
	if err := json.Unmarshal(raw, &s); err == nil && s != "" {
		return []rawBlock{{Type: "text", Text: s}}
	}
	return nil
}

// toolResultText reads a tool_result block's own content field, which has
// the same string-or-blocks shape as a message's content -- transcript.py's
// _text_of joins multiple text blocks with a space; this does the same.
func toolResultText(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var s string
	if err := json.Unmarshal(raw, &s); err == nil {
		return s
	}
	var blocks []struct {
		Text string `json:"text"`
	}
	if err := json.Unmarshal(raw, &blocks); err == nil {
		parts := make([]string, 0, len(blocks))
		for _, b := range blocks {
			if b.Text != "" {
				parts = append(parts, b.Text)
			}
		}
		return strings.Join(parts, " ")
	}
	return ""
}
